File: android/app/src/main/java/com/hcjeong/forestix/common/BreastHeightMarkerLayout.kt
package com.hcjeong.forestix.common

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/// Screen coordinates are dp, matching iOS points. Off-screen heights are
/// hidden rather than clamped to a different position on the screen.
data class BreastHeightMarkerLayout(
    val leftTick: ClosedFloatingPointRange<Float>?,
    val rightTick: ClosedFloatingPointRange<Float>?,
    val labelCenterX: Float?,
    val y: Float,
) {
    companion object {
        const val LABEL_WIDTH = 64f
        const val LABEL_HEIGHT = 22f

        fun make(x: Float, y: Float, width: Float, height: Float,
                 stemLeft: Float? = null, stemRight: Float? = null): BreastHeightMarkerLayout? {
            if (!listOf(x, y, width, height).all { it.isFinite() } ||
                width <= 24 || height <= 24 || x < 0 || x > width ||
                y < 12 || y > height - 12) return null
            var left = x - 24
            var right = x + 24
            if (stemLeft != null && stemRight != null && stemLeft.isFinite() &&
                stemRight.isFinite() && stemLeft < stemRight && x in stemLeft..stemRight &&
                abs(y - height / 2) <= 44) {
                left = min(left, stemLeft - 8)
                right = max(right, stemRight + 8)
            }
            val l = if (left - 12 >= 12) (left - 12)..left else null
            val r = if (right + 12 <= width - 12) right..(right + 12) else null
            if (l == null && r == null) return null
            val label = when {
                r != null && r.endInclusive + 6 + LABEL_WIDTH <= width - 12 ->
                    r.endInclusive + 6 + LABEL_WIDTH / 2
                l != null && l.start - 6 - LABEL_WIDTH >= 12 ->
                    l.start - 6 - LABEL_WIDTH / 2
                else -> null
            }
            return BreastHeightMarkerLayout(l, r, label, y)
        }
    }
}

File: TimberCruisingApp/Common/BreastHeightMarkerLayout.swift
import Foundation

/// Screen-space chrome around a projected world point, in points (dp on
/// Android). Never moves the height to keep an off-screen marker visible.
public struct BreastHeightMarkerLayout {
    public var leftTick: ClosedRange<Double>?
    public var rightTick: ClosedRange<Double>?
    public var labelCenterX: Double?
    public let y: Double
    public static let labelWidth: Double = 64
    public static let labelHeight: Double = 22

    public init?(x: Double, y: Double, width: Double, height: Double,
                 stemLeft: Double? = nil, stemRight: Double? = nil) {
        guard [x, y, width, height].allSatisfy({ $0.isFinite }),
              width > 24, height > 24, x >= 0, x <= width,
              y >= 12, y <= height - 12 else { return nil }
        self.y = y
        var left = x - 24
        var right = x + 24
        // Near the measurement row, keep ticks and text outside the bracket.
        if let l = stemLeft, let r = stemRight, l.isFinite, r.isFinite,
           l < r, x >= l, x <= r, abs(y - height / 2) <= 44 {
            left = min(left, l - 8)
            right = max(right, r + 8)
        }
        leftTick = left - 12 >= 12 ? (left - 12)...left : nil
        rightTick = right + 12 <= width - 12 ? right...(right + 12) : nil
        guard leftTick != nil || rightTick != nil else { return nil }
        if let tick = rightTick, tick.upperBound + 6 + Self.labelWidth <= width - 12 {
            labelCenterX = tick.upperBound + 6 + Self.labelWidth / 2
        } else if let tick = leftTick, tick.lowerBound - 6 - Self.labelWidth >= 12 {
            labelCenterX = tick.lowerBound - 6 - Self.labelWidth / 2
        }
    }
}
